### ISSUE TITLE
Implement ENTROPY-SGD

### ISSUE THREAD

**Issue by bottydim** [fchollet/keras#6563]
A recently proposed method for faster and more robust(better generalization) optimization algorithm:
https://openreview.net/pdf?id=B1YfAfcgl

**Comment by gokceneraslan** [comment]
Nice idea. Updating weights in every `L`th batch seems to be the only technical challenge.

**Comment by bottydim** [comment]
I believe L refers to the number of Langevin iterations, which is a different way to think about batches, I suppose. The code is already available in Lua thanks to the courtesy of Pratik Chaudhari https://github.com/ucla-vision/entropy-sgd

**Comment by kzk2000** [comment]
For some deep math reasons as to why Entropy-SGD is attractive, this paper might also be of interest: https://arxiv.org/pdf/1704.04932v1.pdf

**Comment by maxleverage** [comment]
The issue here might be doing this within an optimizer class as you'll need to get L gradient updates on the same batch of data.

**Comment by gokceneraslan** [comment]
I'm working on it, we can test and improve it together when it's ready.

**Comment by maxleverage** [comment]
Here's the PyTorch version: https://github.com/ucla-vision/entropy-sgd/blob/master/python/optim.py

Looks like we need a closure() type function

**Comment by gokceneraslan** [comment]
I think it's possible to implement it using the `get_updates()` method of the Optimizer class. Implementing a closure or another mechanism which can process multiple batches from within `get_updates()` would diverge from the "Keras way" a lot.

Here is what I have in mind instead, first defining a conditional update op:

```python
cond_update = lambda cond, var, value: K.switch(cond,
                                                lambda: K.update(var, value),
                                                lambda: K.identity(var))
```
Conditional updates are very tricky on TF backend, see [this](https://github.com/fchollet/keras/issues/6621) for more details.

Then the idea is to implement `if iter % L == 0:` and `if iter % L == (L-1):` conditions symbolically, where `iter` is the counter of batch iterations and L is the total number of Langevin iterations. This will force `get_updates()` to behave more statefully based on the value of the iteration counter:

```python
# for the updates that will be performed in every 0th, Lth, (2L)th etc.
sgld_init_cond = K.equal(K.mod(self.iterations, L), 0)
# for the updates that will be performed in every (L-1)th, (2L-1)th etc.
sgld_final_cond = K.equal(K.mod(self.iterations, L), (L-1))
```

`K.mod` is trivial to implement. Now, it's easy to implement for example scoping:

```python
self.updates.append(cond_update(sgld_init_cond,
                                self.sgld_g,
                                self.gamma*((1+self.scoping)**self.iterations)))
```
If `L` is 10, for example and self.iterations counts from 0, then this will only be performed in every 0th, Lth, 2Lth etc. iterations.

After the usual SGD and SGLD steps, weights can be updated simply by:

```python
self.updates.append(cond_update, sgld_final_cond, p, new_p)
```
which will only update the weights in every (L-1)th, (2L-1)th etc. steps.

**Comment by gokceneraslan** [comment]
So that's my answer to #6631 basically.

**Comment by maxleverage** [comment]
Ok thanks! Very helpful.

**Comment by maxleverage** [comment]
Hi guys, I had a crack at the get_updates func on the weekend. Note that momentum and Nester's momentum isn't implemented yet. TF funcs are used directly vs. from the backend for now, but the mod and unit Normal pdf are trivial to implement in the backend.

Would really appreciate a review and feedback.

`    def get_updates(self, params, constraints, loss):
        # Define empty list for update
        self.updates = []

        lr = self.lr
        if self.initial_decay > 0:
            lr *= (1. / (1. + self.decay * self.iterations))
            self.updates.append(K.update_add(self.iterations, 1))

        # Momentum
        shapes = [K.get_variable_shape(p) for p in params]
        moments = [K.zeros(shape) for shape in shapes]
        self.weights = [self.iterations] + moments

        # Use t for tf logical conds, starting from 1. (easier for mod logical control)
        t = self.iterations + 1
        # Update scoping
        g_t = K.pow(tf.multiply(self.g_0, (1 + self.g_1)), t)

        def fn_langevin_1():
            # Initialise mu
            mu = params
            # Save SGD parameters prior to running L iter to estimate local entropy
            self.params_sgd = params

            # Get Langevin gradients
            # In first iter of Langevin loop, x' is the same as SGD params
            langevin_grad = self.get_gradients(loss, params)

            # Update x' and mu
            # Params SGD represents original error surface parameters
            for x_dash, p, m, g in zip(params, self.params_sgd, mu, langevin_grad):
                diff = - tf.multiply(g_t, tf.subtract(p, x_dash))
                delta_x_dash = g + diff
                # Unit Gaussian noise scalar to scale thermal noise
                g_noise = tf.random_normal([1], mean=0.0, stddev=1.0)
                # Update x_dash_new
                x_dash_new = x_dash - self.sgld_lr * delta_x_dash + tf.multiply(tf.sqrt(self.sgld_lr), g_noise) * self.thermal_noise
                # Calculate new mu
                m_new = (1. - self.alpha) * m + self.alpha * x_dash_new
                # Update mu
                K.update(m, m_new)
                # Update x_dash with x_dash_new and append as new params for next mini-batch
                self.updates.append(K.update(x_dash, x_dash_new))
            # Update iter
            K.update_add(self.iterations, 1)
            # Save current mu for next update
            self.mu = mu
            return x_dash


        # Langevin iteration function for Langevin iterations != 1
        def fn_langevin():
            # This function is called when tf.mod(t, self.L) != 0 and != 1 (first iter where x' and mu need to be assigned)
            # Instantiate mu
            mu = self.mu
            # Get gradients to estimate local entropy
            langevin_grad = self.get_gradients(loss, params)
            # Update x' and mu
            for x_dash, p, m, g in zip(params, self.params_sgd, mu, langevin_grad):
                diff = - g_t * tf.subtract(p, x_dash)
                delta_x_dash = g + diff
                # Unit Gaussian noise scalar to scale thermal noise
                g_noise = tf.random_normal([1], mean=0.0, stddev=1.0)
                # Update x_dash_new
                x_dash_new = x_dash - self.sgld_lr * delta_x_dash + tf.multiply(tf.sqrt(self.sgld_lr), g_noise) * self.thermal_noise
                # Update m_new
                m_new = (1. - self.alpha) * m + self.alpha * x_dash_new
                # Update mu
                K.update(m, m_new)
                # Update x_dash with x_dash_new
                self.updates.append(K.update(x_dash, x_dash_new))
            # Update iter
            K.update_add(self.iterations, 1)
            # Update mu
            self.mu = mu
            return x_dash


        # Update for SGD in final Langevin iteration
        def fn_sgd():
            # SGD update function called on final iter of Langevin loop
            # Instantiate saved mu from last iter
            mu = self.mu
            # Get langevin gradients 
            langevin_grad = self.get_gradients(loss, params)
            # Run Langevin updates
            for x_dash, p, m, g in zip(params, self.params_sgd, mu, langevin_grad):
                diff = - g_t * tf.subtract(p, x_dash)
                delta_x_dash = g + diff
                # Unit Gaussian noise scalar to scale thermal noise
                g_noise = tf.random_normal([1], mean=0.0, stddev=1.0)
                # Update x_dash_new
                x_dash_new = x_dash - self.sgld_lr * delta_x_dash + tf.multiply(tf.sqrt(self.sgld_lr), g_noise) * self.thermal_noise
                # Update mu with exponential averaging
                m_new = (1. - self.alpha) * m + self.alpha * x_dash_new
                # Update mu
                K.update(m, m_new)
                # Update x_dash params to update SGD params in error space
                K.update(x_dash, x_dash_new)

            # Update SGD parameters corresponding to original error L iter ago
            for p, m in zip(self.params_sgd, mu):
                # Update original SGD parameters
                new_p = p - (self.lr * g_t) * tf.subtract(p, m)

                # Apply constraints
                if p in constraints:
                    c = constraints[p]
                    new_p = c(new_p)

                # Update p with new_p
                self.updates.append(K.update(p, new_p))

            # Update iter
            K.update_add(self.iterations, 1)
            # Update mu
            self.mu = mu
            return p 

        # Called to return another tf.cond on fn_langevin_1 or fn_langevin
        def tf_cond_elif():
            return tf.cond(condition_langevin_init, lambda: fn_langevin_1(), lambda: fn_langevin())

        # Run condition
        # condition_langevin_init is true when it is the first iteration of Langevin dynamics (required for assignment)
        # condition_langevin is true when iter is less than Langevin iterations
        # fn_sgd is called on final Langevin iter to update original SGD parameters
        condition_langevin_init = tf.equal(tf.mod(t, self.L), 1)
        condition_langevin = tf.not_equal(tf.mod(t, self.L), 0)

        # Execute nested tensorflow conditions
        # Updates happen inside lambda functions, cond_res result is not used
        cond_res = tf.cond(condition_langevin, lambda: tf_cond_elif(), lambda: fn_sgd())
        
        # Return updates
        return self.updates`

**Comment by javiercorrea** [comment]
Anyone still working on this?

I try to implement it myself, but I'm not very comfortable with Keras inner workings yet.

Here's my implementation thus far:

```
from keras.optimizers import Optimizer
from keras.legacy import interfaces
import keras.backend as K

import tensorflow as tf
K.mod = lambda x,y: tf.mod(x,y)

class EntropySGD(Optimizer):
    def __init__(self, lr=0.1, L=5, gamma=1E-1, epsilon=1E-4, alpha=0.75,
                 inner_momentum=0., outer_momentum=0., decay=0., nesterov=False, **kwargs):
        super(EntropySGD, self).__init__(**kwargs)
        with K.name_scope(self.__class__.__name__):
            self.iterations = K.variable(0, dtype='int64', name='iterations')
            self.lr = K.variable(lr, name='lr')
            self.L = K.variable(L, name='L', dtype='int64', name='L')
            self.gamma = K.variable(gamma, name='gamma')
            self.epsilon = K.variable(epsilon, name='epsilon')
            self.alpha = K.variable(alpha, name='alpha')
            self.inner_momentum = K.variable(inner_momentum, name='inner_momentum')
            self.outer_momentum = K.variable(outer_momentum, name='outer_momentum')
            self.decay = K.variable(decay, name='decay')
        self.initial_decay = decay
        self.nesterov = nesterov

    @interfaces.legacy_get_updates_support
    def get_updates(self, loss, params):
        grads = self.get_gradients(loss, params)
        self.updates = [K.update_add(self.iterations, 1)]

        lr = self.lr
        if self.initial_decay > 0:
            lr *= (1. / (1. + self.decay * K.cast(self.iterations,
                                                  K.dtype(self.decay))))
        # momentum
        shapes = [K.int_shape(p) for p in params]
        inner_moments = [K.zeros(shape) for shape in shapes]
        outer_moments = [K.zeros(shape) for shape in shapes]
        self.weights = [self.iterations] + inner_moments + outer_moments
        mu_s = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
        pp_s = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]

        zippped = zip(params, grads, inner_moments, outer_moments, mu_s, pp_s)
        for k, (p, g, m_i, m_o, mu, pp) in enumerate(zippped):
            new_langevin = K.equal(K.mod(self.iterations, self.L), 0)
            self.updates.append(K.switch(new_langevin,
                                        K.update(mu, p),
                                        K.identity(mu)))
            self.updates.append(K.switch(new_langevin,
                                        K.update(pp, p),
                                        K.identity(pp)))
            self.updates.append(K.switch(new_langevin,
                                        K.update(m_i, K.zeros(shapes[k])),
                                        K.identity(m_i)))

            v_i = self.inner_momentum*m_i - lr*(g + self.gamma*(p-pp))
            self.updates.append(K.update(m_i, v_i))

            noise = K.random_normal_variable(shape=K.int_shape(pp),
                                             mean=0.0, scale=1.0)
            if self.nesterov:
                new_pp = pp + self.inner_momentum*v_i - lr*(g + self.gamma*(p-pp))\
                        + K.sqrt(lr)*self.epsilon*noise
            else:
                new_pp = pp + v_i + K.sqrt(lr)*self.epsilon*noise
            new_mu = (1.0 - self.alpha)*mu + self.alpha*pp

            self.updates.append(K.update(pp, new_pp))
            self.updates.append(K.update(mu, new_mu))

            v_o = self.outer_momentum*m_o - lr*self.gamma*(pp - mu)
            self.updates.append(K.update(m_o, v_o))

            if self.nesterov:
                new_p = p + self.outer_momentum*v_o - lr*self.gamma*(p - mu)
            else:
                new_p = p + v_o

            update_params = K.equal(K.mod(self.iterations, self.L-1), 0)
            self.updates.append(K.switch(update_params,
                                         K.update(p, new_p),
                                         K.identity(p)))
        return self.updates

    def get_config(self):
        config = {'lr': float(K.get_value(self.lr)),
                  'L': int(K.get_value(self.L)),
                  'inner_momentum': K.get_value(self.inner_momentum),
                  'outer_momentum': K.get_value(self.outer_momentum),
                  'gamma': float(K.get_value(self.gamma)),
                  'epsilon': float(K.get_value(self.epsilon)),
                  'alpha': float(K.get_value(self.alpha)),
                  'decay': float(K.get_value(self.decay)),
                  'nesterov': self.nesterov}
        base_config = super(EntropySGD, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))
```

**Comment by skudekar** [comment]
Hi, 

> Anyone still working on this?
> 
> I try to implement it myself, but I'm not very comfortable with Keras inner workings yet.
> 
> Here's my implementation thus far (although it doesn't seem to be working):
> 
> ```
> from keras.optimizers import Optimizer
> from keras.legacy import interfaces
> import keras.backend as K
> 
> import tensorflow as tf
> K.mod = lambda x,y: tf.mod(x,y)
> K.floordiv = lambda x,y: tf.floordiv(x,y)
> 
> class EntropySGD(Optimizer):
>     def __init__(self, lr=1E-2, inner_lr=1E-1, L=10, gamma0=0.0, gamma1=1E-1, epsilon=1E-3, alpha=0.75,
>                  inner_momentum=0., outer_momentum=0., decay=0., nesterov=False, **kwargs):
>         super(EntropySGD, self).__init__(**kwargs)
>         with K.name_scope(self.__class__.__name__):
>             self.iterations = K.variable(0, dtype='int64', name='iterations')
>             self.inner_lr = K.variable(inner_lr, name='inner_lr')
>             self.lr = K.variable(lr, name='lr')
>             self.L = K.variable(L, name='L', dtype='int64')
>             self.gamma0 = K.variable(gamma0, name='gamma0')
>             self.gamma1 = K.variable(gamma1, name='gamma1')
>             self.epsilon = K.variable(epsilon, name='epsilon')
>             self.alpha = K.variable(alpha, name='alpha')
>             self.inner_momentum = K.variable(inner_momentum, name='inner_momentum')
>             self.outer_momentum = K.variable(outer_momentum, name='outer_momentum')
>             self.decay = K.variable(decay, name='decay')
>         self.initial_decay = decay
>         self.nesterov = nesterov
> 
>     @interfaces.legacy_get_updates_support
>     def get_updates(self, loss, params):
>         grads = self.get_gradients(loss, params)
>         self.updates = [K.update_add(self.iterations, 1)]
> 
>         lr = self.lr
>         t = K.cast(K.floordiv(self.iterations, self.L), K.dtype(self.gamma1))
>         if self.initial_decay > 0:
>             lr *= (1. / (1. + self.decay * t))
>         gamma = self.gamma0*K.pow(1.0 + self.gamma1, t)
> 
>         # momentum
>         shapes = [K.int_shape(p) for p in params]
>         inner_moments = [K.zeros(shape) for shape in shapes]
>         outer_moments = [K.zeros(shape) for shape in shapes]
>         self.weights = [self.iterations] + inner_moments + outer_moments
>         mu_s = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
>         pp_s = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
> 
>         new_langevin = K.equal(K.mod(self.iterations, self.L), 0)
>         update_params = K.equal(K.mod(self.iterations, self.L), self.L-1)
> 
>         zippped = zip(params, grads, inner_moments, outer_moments, mu_s, pp_s)
>         for k, (p, g, m_i, m_o, mu, pp) in enumerate(zippped):
>             mu_ = K.switch(new_langevin,
>                            lambda: K.identity(p),
>                            lambda: K.identity(mu))
>             pp_ = K.switch(new_langevin,
>                            lambda: K.identity(p),
>                            lambda: K.identity(pp))
>             m_i_ = K.switch(new_langevin,
>                             lambda: K.constant(0.0, shape=shapes[k]),
>                             lambda: K.identity(m_i))
> 
>             # Inner optimization SGLD
>             dpp = g - gamma*(p - pp_)
>             v_i = self.inner_momentum*m_i_ - self.inner_lr*dpp
>             self.updates.append(K.update(m_i, v_i))
> 
>             noise = K.random_normal(shape=shapes[k],
>                                     mean=0.0,
>                                     stddev=K.sqrt(self.inner_lr)*self.epsilon)
>             if self.nesterov:
>                 new_pp = pp_ + self.inner_momentum*v_i\
>                         - self.inner_lr*dpp + noise
>             else:
>                 new_pp = pp_ + v_i + noise
> 
>             new_mu = (1.0 - self.alpha)*mu_ + self.alpha*new_pp
>             self.updates.append(K.update(pp, new_pp))
>             self.updates.append(K.update(mu, new_mu))
> 
>             # Outer optimization SGD
>             dp = p - new_mu
>             v_o = self.outer_momentum*m_o - lr*dp
>             self.updates.append(K.update(m_o, v_o))
> 
>             if self.nesterov:
>                 new_p = p + self.outer_momentum*v_o - lr*dp
>             else:
>                 new_p = p + v_o
> 
>             self.updates.append(K.switch(update_params,
>                                          lambda: K.update(p, new_p),
>                                          lambda: K.identity(p)))
>         return self.updates
> 
>     def get_config(self):
>         config = {'lr': float(K.get_value(self.lr)),
>                   'inner_lr': float(K.get_value(self.lr)),
>                   'L': int(K.get_value(self.L)),
>                   'inner_momentum': K.get_value(self.inner_momentum),
>                   'outer_momentum': K.get_value(self.outer_momentum),
>                   'gamma0': float(K.get_value(self.gamma0)),
>                   'gamma1': float(K.get_value(self.gamma1)),
>                   'epsilon': float(K.get_value(self.epsilon)),
>                   'alpha': float(K.get_value(self.alpha)),
>                   'decay': float(K.get_value(self.decay)),
>                   'nesterov': self.nesterov}
>         base_config = super(EntropySGD, self).get_config()
>         return dict(list(base_config.items()) + list(config.items()))
> ```

Hi, did you manage to get this working? I tried to implement the code above for MNIST and it seems to work for L = 5,10. But at L=20 the accuracy in training is very low. Is this what you also experience? thanks for any help!!
